distributed/protocol/serialize.py

```diff
@@ -71,29 +74,116 @@ def pickle_loads(header, frames):
+        _cached_allowed_modules[name] = importlib.import_module(name)
+        return _cached_allowed_modules[name]
+    else:
+        raise RuntimeError(
+            f"Importing {repr(name)} is not allowed, please add it to the list of "
+            "allowed modules the scheduler can import via the "
+            "distributed.scheduler.allowed-imports configuration setting."
+        )
+
+
 def msgpack_decode_default(obj):
     """
-    Custom packer/unpacker for msgpack to support Enums
+    Custom packer/unpacker for msgpack
     """
     if "__Enum__" in obj:
-        mod = importlib.import_module(obj["__module__"])
-        enum_type = getattr(mod, obj["__name__"])
-        obj = getattr(enum_type, obj["name"])
+        mod = import_allowed_module(obj["__module__"])
+        typ = getattr(mod, obj["__name__"])
+        return getattr(typ, obj["name"])
+
+    if "__Set__" in obj:
+        return set(obj["as-list"])
+
+    if "__SubgraphCallable__" in obj:
+        mod = import_allowed_module(obj["__module__"])
+        typ = getattr(mod, obj["__name__"])
+        return typ(*obj["args"])
+
+    if "__Layer__" in obj:
+        obj_name = obj["__name__"]
+        if obj_name == "BasicLayer":
+            # The default implemention of Layer returns a BasicLayer, which might
+            # not be defined in `mod` therefore we import it explicitly here
+            typ = BasicLayer
+        else:
+            mod = import_allowed_module(obj["__module__"])
+            typ = getattr(mod, obj["__name__"])
+        return typ(*obj["args"])
+
+    if "__HighLevelGraph__" in obj:
+        return HighLevelGraph(
+            obj["layers"],
+            obj["dependencies"],
+        )
+
+    if "__Serialized__" in obj:
+        # Notice, the data here is marked a Serialized rather than deserialized. This
+        # is because deserialization requires Pickle which the Scheduler cannot run
+        # because of security reasons.
+        # By marking it Serialized, the data is passed through to the workers that
+        # eventually will deserialize it.
+        return Serialized(*obj["data"])
+
     return obj
 
 
 def msgpack_encode_default(obj):
     """
-    Custom packer/unpacker for msgpack to support Enums
+    Custom packer/unpacker for msgpack
     """
 
+    if isinstance(obj, Serialize):
+        return {"__Serialized__": True, "data": serialize(obj.data)}
+
     if isinstance(obj, Enum):
         return {
             "__Enum__": True,
             "name": obj.name,
             "__module__": obj.__module__,
             "__name__": type(obj).__name__,
         }
+
+    if isinstance(obj, set):
+        return {"__Set__": True, "as-list": list(obj)}
+
+    if isinstance(obj, SubgraphCallable):
+        cls, args = obj.__reduce__()
+        return {
+            "__SubgraphCallable__": True,
+            "__module__": obj.__module__,
+            "__name__": cls.__name__,
+            "args": args,
+        }
```

Comment:
We might not need to special-case this in the Scheduler.  My guess/hope is that the scheduler won't need to think too much about the payload within a Blockwise or other layer.  Instead it will just have to know that there is a binary payload of bytes that it needs to include in every task.

I haven't gone deeply into this problem though, and so I may not fully understand what is going on here.